Windows MSI Packaging

rozmansi opened the pull request:
This PR adds MSI packaging side-by-side with NSIS packaging.

## Quick Instructions

### MSI Packaging
See detailed instructions in https://github.com/rozmansi/openvpn-build/blob/feature/windows-msi/windows-msi/README.rst.

## Binaries for Initial Impression
Compiled, signed and MSI packaged OpenVPN is available at https://github.com/rozmansi/openvpn/releases/tag/v2.4.6m. This is for informational purposes, to get a quick overview of setup experience.

## Final Note
Please mind that this PR starts with a packaging of OpenVPN 2.4.6 version instead of 2.5._git on purpose. We need to have previous version of the package to test the upgrading process before 2.5 is published.

mattock commented on windows-msi/README.rst:
```diff
@@ -0,0 +1,99 @@
+  set a description in the MSI signature accurately describing the package
+  content.
+
+- MSI packages do not support multiple signatures. Sign them only once using
```
We (=OpenVPN project) do not officially support Windows Vista anymore. So SHA-256 signature should probably be the recommendation.

rozmansi replied:
True.

It makes sense to update the document then since it also mentions double signing of EXE/DLL binaries. If Vista is no longer supported, SHA-1+SHA-256 double signing is a thing of past.

mattock replied:
You're referring to documentation on Trac, right? I had a quick look and the signing part requires a major overhaul. I created an internal JIRA ticket assigned to myself to fix that after the next tap-windows6 release.

rozmansi replied:
Resolved in a9b1331c763d388aaa51b72d9e12deb5bcbfe0d9.

mattock commented on windows-msi/bookmarks/howto.url:
```diff
@@ -0,0 +1,4 @@
+[DEFAULT]
+BASEURL=https://openvpn.net/howto.html
+[InternetShortcut]
+URL=https://openvpn.net/howto.html
```
This may be a silly question, but why two sections with almost identical content?

rozmansi replied:
Ask Microsoft! :) This is how Internet Explorer authors the .url files. The .url file launches default web browser when opened by the shell. In other words: it's an URL version of a .lnk shortcuts.

Frankly, I didn't take time to test if the shell is happy with `URL` in `[InternetShortcut]` section alone. Wouldn't be surprised to see different behaviour on different Windows versions. So it seemed least collateral damage to leave both URL's in the file vs. testing all Windows versions and reverse engineering the file structure in the absence of official documentation.

Unfortunately, there is no official MSDN documentation on the .url files. The best description you can find online is [here](https://www.codeproject.com/articles/22267/internet-explorer-favorites-deconstructed). Reverse engineered by somebody.

mattock replied:
Ok. I don't mind, I was just wondering why :smile:.

mattock commented on windows-msi/boot.wsf.in:
```diff
@@ -0,0 +1,112 @@
+@PRODUCT_NAME@-install-@PRODUCT_VERSION@.exe ADDLOCAL=driver,util,sdk
+   Install "driver", "util" and "sdk" features.
+
+@PRODUCT_NAME@-install-@PRODUCT_VERSION@.exe INSTALLDIR=&lt;folder&gt;
```
Will @PRODUCT_VERSION@ include the package version? For example, with NSIS, we have a version string like 2.4.6-I601, where I601 represents the package version. We increment the package version whenever dependencies (mainly openssl) need to be updated.

rozmansi replied:
Unfortunately, no. But, I can add that to the filename.

For package versioning with MSI we shall use:
- same product version again
- same product code GUID
- change the package code GUID (which is auto-generated by WiX, therefore all we need is to rebuild the MSI packages)

Source: https://docs.microsoft.com/en-us/windows/desktop/Msi/productversion

I have added a bullet on my TODO list to test MSI small updates (same product, different package) with OpenVPN packages explicitly.

rozmansi replied:
> "Houston, we have a problem..."

I have introduced the package version support and played with it for a while.

1. MSI will refuse to install a .msi package with the same product version, product code, and different package code over the existing one. MSI considers this a _small update_ and only way to install it is to create and deploy a .msp patch. This will complicate our deployment flow: you'll need to publish all .msp patches for Group Policy admins, I'll need to add .msp patches to SFX .exe installer payload and update the bootstrapper appropriately...

2. Next option is so-called _minor update_: we keep the product code but change the product version. An example of a minor update is a service pack. We can deploy it as a .msp patch or a full-fledged .msi package. The later option suits us best. But (I hate this word), the MSI honors only first three version fields. "2.4.6" or "2.4.6.1" is all the same product version for the MSI. Which won't fly with existing OpenVPN versioning scheme. One possible workaround is we shift version left by one. But only for MSI's `ProductVersion` property. Filenames and MSI GUI would still display "OpenVPN 2.4.6-1". Imagine shifting "openvpn-2.4.6-1" to "openvpn2-4.6.1". It is still possible to author OpenVPN3 MSI as an upgrade to OpenVPN2 MSI even with this change.
The only negative side effect is the legacy Programs and Features app from Control Panel will display product version as 4.6.1 in the package info.

I suppose we can't easily just quickly release 2.4.7 should some dependency require an immediate CVE deployment, right?

selvanair replied:
Instead of shifting off the major version (2), what about combining release+build: say 2.4.6.0 is treated as 2.4.60, 2.4.6.2 as 2.4.62 etc. As long as we wont have more than 10 (or 16 if converted as hex) "minor" updates, this would work. Not ideal but better than version becoming 4.6.1 etc.

mattock replied:
We've never had more than 3-4 NSIS installer releases. So ten is more than adequate. It will will be confusing, but probably less confusing than having OpenVPN 4.6.1 show up on the system :smile:.

rozmansi replied:
Selva, that's very clever, thanks - I shall go for it. Samuli, the product version as authored by `ProductVersion` property in MSI is not displayed a lot.

Quoting the MSDN:

>
>
> The first field is the major version and has a maximum value of 255. The second field is the minor version and has a maximum value of 255. The third field is called the build version or the update version and has a maximum value of 65,535.

This means we have plenty of space in the third version field. So, let's use a factor of 100:
2.4.600, 2.4.601 etc.

rozmansi replied:
Resolved in db3deac24002e327a474e458ed9f1dfdbb25788e. However, I still need to thoroughly test minor updates before going live.

mattock commented on windows-msi/version.m4:
```diff
@@ -0,0 +1,37 @@
+
+dnl The upgrade GUIDs MUST persist for all versions of the same product line.
+dnl Please use own upgrade GUIDs when deploying a non-official OpenVPN release.
+define([PRODUCT_UPGRADE_GUID_x86],   [{1195A47B-A37A-4055-9D34-B7A691F7E97B}])
```
How were/are these GUIDs created?

rozmansi replied:
`uuidgen | tr [a-f] [A-F]`

I just picked some random to get started. Any unique hexadecimal number is OK. Old MSI versions required all uppercase GUIDs and I stick to that recommendation.

`PRODUCT_UPGRADE_GUID_...` are called upgrade code(s) and are used by MSI to follow the product line across versions, 3rd party products to detect OpenVPN is installed... The upgrade code is rarely changed - it is _the_ product identifier. However, I choose not to hard-code them to allow 3rd party vendors to package their own "OpenVPN" product line.

Use of platform-specific upgrade codes teaches Group Policy Management Editor not to mix 32 and 64-bit packages when updating. Example: MS Office 2016 has something like `{00160000-008C-0000-0000-0000000FF1CE}` for 32-bit version and `{00160000-008C-0000-1000-0000000FF1CE}` for 64-bit.

If you want some fancy ones to mark the official line of 32 and 64-bit OpenVPN MSI packages, it is time to set them now.

selvanair commented on windows-msi/msi.wxs:
```diff
@@ -925,7 +925,7 @@
                 Title="$(var.PRODUCT_NAME) Service"
                 Description="Service wrappers"
-                Level="2"
+                Level="3"
```
Currently we install this by default, dont we?

rozmansi replied:
True, but it would also mean I need to set the `openvpnserv2.exe` service start to OnDemand/Manual - to keep consistent experience with NSIS installations.

But, I am afraid users needing this service will always get this service reset to Manual start after updating OpenVPN.

There is a subtle difference between NSIS and MSI... In MSI we could query the configured service start type in registry like NSIS does. Unfortunately, we cannot use that property in MSI's `ServiceInstall` table, as the `StartType` column is `DoubleInteger`, not `Formatted` type. :(

In order to preserve the service start configuration over updates we would need quite some hacking (runtime MSI database manipulation) and/or some additional custom actions.

I made a note to research whether MSI resets the service start type when the `openvpnserv2.exe` component is updated.

rozmansi replied:
Update: `openvpnserv2.exe` service is already set to OnDemand start in MSI package. I have set it to install by default now to provide a more seamless transition from NSIS installers.

selvanair commented on windows-msi/msi.wxs:
```diff
@@ -387,10 +568,68 @@
                         <File Name="license.txt" Source="!(bindpath.build)$(var.PRODUCT_PLATFORM)\license.txt"/>
 
                         <!-- Clean-up NSIS installation. -->
```
This is not the right place to comment on this but finding the right commit in git PRs is hard when the patch big and commits numerous.. Or I don't know how to tame it on a browser -- much easier with git and vi :)

Anyway, the registry entry disable_save_passwords is DWORD but here it is read back as raw -- that will set it to "#0" or "#1" and when written back as integer it will become REG_SZ (WiX magic, I guess).

An option is to set its default to "#0", read as raw, and then write as string which will automatically become DWORD because of the # prefix  --- I suppose. Or you may know a better way.

rozmansi replied:
You're right. Missed that. I checked the final MSI and the WiX indeed authored the `Value` column of `Registry` table to `#[DISABLESAVEPWD]`, which will render incorrectly to `##0` or `##1` on updates.

I see no better to address this other than you suggested.

selvanair commented:
And one more thing: If the service or GUI is running is it stopped to allow for file updates or does the user have to intervene? Quitting the GUI is easy for the user, but at least for the service we should stop and restart as/when required. Couldn't figure how/whether that's done.

rozmansi commented on windows-msi/Makefile:
```diff
@@ -0,0 +1,229 @@
+	"$(SOURCE_DIR)\bin\openvpnserv.exe" \
+	"$(OPENVPNSERV2_DIR)\openvpnserv2.exe" \
+	"$(BUILD_DIR)\README-config.txt" \
+	"..\windows-nsis\INSTALL-win32.txt" \
```
True. However, I can't just simply fork it to the `windows-msi` folder. Its content needs to be updated.

I suppose it would be best to update and move the building instructions to a Wiki page, and link to it from the `INSTALL-win32.txt` document. This document is installed to end-user computers and should focus on Windows-specific OpenVPN setup.

mattock replied:
@rozmansi I think we could move INSTALL-win32.txt to the root of the repository and, as you suggest, move the actual build instructions (NSIS/MSI) to the Trac wiki. The percentage of Windows users who actually cross-compile or package OpenVPN is so small that showing the instructions to everyone is rather pointless.

rozmansi replied:
Fixed in b127c568a04ca3f1ab3df430d1224dad489e47d2.

rozmansi commented:
> 
> 
> And one more thing: If the service or GUI is running is it stopped to allow for file updates or does the user have to intervene? Quitting the GUI is easy for the user, but at least for the service we should stop and restart as/when required. Couldn't figure how/whether that's done.

Services are authored to stop on component uninstall and start on component install. (But, I'll double check that.)

The applications (i.e. `openvpn-gui.exe`) does not. Fortunately, the MSI has files-in-use detection out-of-the-box. The user will be prompted that an "OpenVPN GUI" application is opened and is preventing some files to update. The user will then be able to close the application her/himself and choose Retry, or choose to proceed with installation and restart computer later. Installer Service saves the files it was unable to update and updates them on next restart.

rozmansi commented:
For a better illustration...

1. Installed NSIS OpenVPN 2.4.6 on a vanilla computer and configured a VPN connection.
2. Connected to VPN.
3. Launched the new MSI install to update.

As expected, I get the following prompt.
![image](https://user-images.githubusercontent.com/6889694/48622272-816c4700-e9a6-11e8-8e6f-33df5d3101e0.png)

(OpenVPN Deamon process is openvpn.exe)

Now I can choose to terminate the VPN connection and close the OpenVPN GUI myself before updating. Or, proceed with the update and restart computer later.

4. Ignore the warning.
5. The update proceeds. When it stops the interactive service, the VPN connection is terminated, OpenVPN GUI pops-up to warn that.
6. After the update both openvpn.exe and openvpn-gui.exe were updated in `C:\Program Files\OpenVPN\bin` (verified by signatures), and a reboot was not required. The OpenVPN GUI is still running in the system tray. This puzzles me, because it works too good to be true.

selvanair commented:
>     3. After the update both openvpn.exe and openvpn-gui.exe were updated in `C:\Program Files\OpenVPN\bin` (verified by signatures), and a reboot was not required. The OpenVPN GUI is still running in the system tray. This puzzles me, because it works too good to be true.

Interesting.. I guess it must be moving the original gui exe to a temp location and let it continue running. Renaming files in use is supported by ntfs (though the file explorer UI may not allow it).  What executable path and name are shown in the running process properties? 

Updating from a 2.4.4 or older NSIS install to 2.4.5 or 6 should show this clearly --- the running GUI will report a version of   11.9.0.0 in its settings->About menu but the newly installed file will show 11.10.0.0. The GUI version did not change between 2.4.5 and 2.4.6.

rozmansi commented:
1. Installed 2.3.11, `openvpn-gui.exe` version 10
2. Configured and established a VPN connection.
3. Launched OpenVPN 2.4.6 MSI update over the VPN connection: the MSI file was located on the share accessible via VPN connection.
4. Update warned about the files-in-use. I clicked ignore.
5. Update finished. No prompt for a restart. Files in `C:\Program Files\OpenVPN` were updated. VPN connection was still established. `openvpn-gui.exe` still running (version 10). The _Image path name_ and _Command line_ of the `openvpn-gui.exe` and `openvpn.exe` processes in the Task Manager did not change.
6. Closing and restarting `openvpn-gui.exe` the version 11.10 showed up.

This confirms @selvanair's assumptions. The update indeed allows the established VPN connections to persist until the processed are recycled. This means very good news for updating remote computers over an OpenVPN connection.

Test machine: up-to-date Windows 8.1 Pro.

Now I need to use even older version, to test the TAP driver update.

selvanair commented:
> I am thinking about challenges if we take hybrid approach: When live VPN connections are detected, user is given a choice: "Do you want to preserve established VPN connections? (Note: You will need to restart OpenVPN later to switch to the new version.)" If there are no live connections or user chooses No for the previous question, the MSI does stop-update-start; otherwise update only.

I think this is okay, but we could be a little more helpful by saying the service also need to be restarted and how to do that. Alternatively just say that a reboot is required for changes to take effect (if service was updated) --- Windows users are used to rebooting anyway :)

@mattock may have better suggestions.

mattock commented:
I think the two options suggested by @rozmansi are for different audiences/use-cases:

1. _Enterprises / upgrading remote nodes managed via VPN links:_ you don't want to loose VPN connectivity during upgrades.
2. _End-users / upgrading the local computer:_ you want to avoid having to manually restart services or reboot after OpenVPN upgrade.

A hybrid approach would be best of course. If it is difficult to implement then let's think about which one of the above we should choose. There are good arguments for both. That said, enterprise sysadmins are able to figure out ways to work around caveats in option 2, whereas typical end users might be caught badly off-guard by 1.

rozmansi commented:
The current implementation of MSI packets _was_ the 1st option already. Because it was the simplest to implement. :)

I have changed the MSI package to:

1. Attempt to stop OpenVPN GUI before commencing `ExecuteAction`. Since this action is authored in the `InstallUISequence`, it'll execute only for interactive MSI sessions. The silent (GPO) installs are not affected. Furthermore, since it launches non-elevated and as the current user, it can send the `WM_CLOSE` signal to modern non-elevated running OpenVPN GUI windows only. Therefore, the MSI will not close the old OpenVPN GUI when updating a prehistoric OpenVPN installation.

2. Stop all OpenVPN services before updating.

3. The interactive service will start back after the update. The legacy and `openvpnserv2.exe` will remain stopped.

4. When the user clicks "Close" on the final MSI page, the OpenVPN GUI is relaunched. Again, this happens only for interactive MSI sessions. Furthermore, it is launched only when the _"Launch on User Logon"_ feature is selected.

One more thing about services: The Microsoft Installer is designed to install, configure and start services in a constant way = update resets everything back to defaults. Stock MSI actions are pretty limited. It is impossible to use a variable for the required service state and startup. Hence, the legacy and the `openvpnserv2.exe` _will_ always stop and reset to a manual start on OpenVPN updates. This is a regression from the NSIS based updates users of `openvpnserv2.exe` service might not fancy.

I am desperately trying to find time to add a new set of custom actions to detect which services were running before the update, what startup type they had, and to start and configure them the same after the update is complete. Much like the NSIS installer does now. It's a lot of work and basically means reinventing the wheel (in other words: replacing entire MSI service management by a custom one).

Unless we _can_ afford to have the `openvpnserv2.exe` users start and set for auto-start the `openvpnserv2.exe` manually after each OpenVPN update. (in other words: _We don't care. If you want it, it's still there, start it yourself._)

BTW, a transform can be authored for those users to have it applied at installation/update time to set the `openvpnserv2.exe` to start on install/update and set its startup type to auto-start. Or, have a batch file to activate the service for them.

Or, we could simply author the MSI to start and set for auto-start the `openvpnserv2.exe` too, but demote this feature to level 3 (not selected by default). Of course, the warning should be displayed: _"If you choose to install this feature all your existing `.ovpn` connections will attempt to connect immediately."_ The feature selection is preserved across the updates. So, users of `openvpnserv2.exe` will need to select this feature on the initial installation only; and those who do not want this service, will absolutely have to _not_ select it to install.

Just thinking aloud...

selvanair commented:
If it wont be too much work, an option is to check whether the service is running and if so set a registry key named autostart_config_dir (only if absent) to the same value as config_dir. Else leave it blank. If that's easier to do, we can change the service code to auto start configs in that directory and let the service start type to auto by default. This check is needed only for the first move (from NSIS to MSI) --we are already doing a some housekeeping jobs during install in that case.

I never liked openvpnserv autostarting all configs in the same central config folder that the GUI also reads. We should have had a separate location for such configs or a way of selecting which configs to autostart. This may be an opportunity to do so.

rozmansi commented:
> 
> 
> If it wont be too much work, an option is to check whether the service is running and if so set a registry key named autostart_config_dir (only if absent) to the same value as config_dir. Else leave it blank. If that's easier to do, we can change the service code to auto start configs in that directory and let the service start type to auto by default. This check is needed only for the first move (from NSIS to MSI) --anyway, we are already doing a some housekeeping jobs during install in that case.

This is not too much work and it would provide a seamless update for `openvpnserv2.exe` users indeed. Of course, it will introduce an inconsistency in folder layout for upgraded vs. fresh installs. I wonder what will cause more head-scratching to sysadmins:
1. Getting a different folder layout for upgraded vs. fresh installs: different `autostart_config_dir` equals different behaviour of `openvpnserv2.exe`.
2. Having to read the _"Breaking Changes in v2.5"_ chapter and moving their auto-start .ovpn files from `C:\Program Files\OpenVPN\config` to `C:\Program Files\OpenVPN\config-autostart`.

In the long term, I think it is cleaner and more consistent we always set `autostart_config_dir` to `C:\Program Files\OpenVPN\config-autostart` (and preserve this registry value on future updates), while instructing users of `openvpnserv2.exe` they need to move their auto-start .ovpn files to the newly introduced folder - or - set the `autostart_config_dir` to `C:\Program Files\OpenVPN\config`.

Another problem with `openvpnserv2.exe` is .NET Framework dependency. I am researching how to disable the `openvpnserv2.exe` feature when .NET Framework is absent _before_ authoring MSI to start this service and set it to auto-start. When MSI installs a service and tries to start it, but the service fails to start, the entire MSI session is considered a failure and is rolled back.

> 
> I never liked openvpn service autostarting all configs in the same central config folder that the GUI also reads. We should have had a separate location for such configs or a way of selecting which configs to autostart. This may be an opportunity to do so.

I totally agree. I vote for the separate config folder approach:
- A similar solution is used on Linux already: `/etc/openvpn/client` and `/etc/openvpn/server`
- It is simpler to implement.

(Note that using two separate config folders means user needs to move/duplicate all external files referenced in .ovpn by the relative path too. We could keep the same config folder but have a .ovpn and .ovpn-auto config files. Unfortunately, Windows Explorer makes it difficult to change the file extensions by default. We could change the `openvpnserv2.exe` to pick only `*-service.ovpn` files, while `openvpn-gui.exe` ignores them. Both proposals are $100 solution for a ¢2 problem.)

mattock commented:
:+1: to having a separate folder for automatically started config files. I'm also not against having separate config directories for client and server configs, as long as there are no implicit assumptions about auto-starting based on type. For example, my wife has an always-on client connection to her workplace, so autostarting is not just a thing for servers. But it could get tricky if we also need to distinguish between client, server and autostart configs.

rozmansi commented:
Dividing the config folder into client/server on Windows would only bring confusion. Heck, I don't even see the rationale behind dividing `openvpn@.service` into `openvpn-client@.service` and `openvpn-server@.service` on Linux.

- `C:\Program Files\OpenVPN\config` + `C:\Program Files\OpenVPN\config-auto`, or
- `C:\Program Files\OpenVPN\config` + `C:\Program Files\OpenVPN\config\service`

...comes to my mind. The later would still allow OpenVPN-GUI to run a .ovpn from `config\service` manually, should a user wish to stop the service and connect manually. While preventing `openvpnserv2.exe` from start connecting immediately when user upgrades with the MSI setup.

selvanair commented:
I agree with not distinguishing server and client configs. Very few people run server on Windows and they can fend for themselves.

As for config-auto vs config\service, I prefer the former -- the latter where the GUI sees these configs brings us back closer to the current state: users may try to start those configs from the GUI, may even succeed (if there are enough tap adapters) and cause the auto-started one to disconnect or the two to usurp each other back and forth.

What I had in mind was something like this (using some key and folder names for concreteness, not saying these are the best names)

```
if regkey autostart_config_dir is present 
    do nothing
else if fresh install
    set reg key autostart_config_dir = C:\ProgramFiles\OpenVPN\config-auto
else if OpenVPNService is Running
    set reg key autostart_config_dir = C:\ProgramFiles\OpenVPN\config
```

Then update the service code to use autostart_config_dir reg key to find the config file locations and ignore if its blank, missing, unreadable, empty etc. Also update NSIS installer to create this reg key if we distribute the updated service with future 2.4.x releases.

The only real downside is what @rozmansi  mentioned: the change or config location for new installs. That's the price to pay for separating the two kinds of configs. 

As for users who want to stop the auto-start, they should either move the configs away or rename the folder or the registry key. Stopping the service is not a good method (except as a stopgap measure), as upgrades will bring the service back on. 

BTW, "auto" may not be the right word to describe these configs -- in the GUI also we can have some configs to "autostart" so the meaning of auto is context-dependent. That said, I don't have a better word -- "service", "daemon", "always-on", "prestarted" ?  -- and am ok with "auto".

mattock commented:
On Linux there was solid reasoning behind separating client and server configs, but I've forgotten the details. It probably had to do with init script / systemd unit file ordering on boot, but I could be mistaken. On Windows I think it would just confuse people. For me config-auto seems reasonable. The pseudo-code by @selvanair looks good.

selvanair commented:
@rozmansi To avoid your concern about folder name depends on upgrade/fresh-install issue, we could copy the contents of config to config-auto when appropriate: key not found but service running, so we add the key and copy the contents of config. Then the reg key value also stays independent of fresh install or upgrade.

Documenting the change becomes easier that way.

rozmansi commented:
@selvanair Can you do the `openvpnserv2.exe` part, please? And I do the MSI part following your recommendations.

selvanair commented:
Looking at openvpnserv2 -- it read both 32 bit and 64 bit registries and starts configs in both (using appropriate exe_path found in each place). Is that what is expected or can we just require 64 bit OpenVPN installation on 64 bit machines?

Else the installer doing 64 bit install will have to also look for existing 32 bit install and update its registry, copy config folder etc..

@mattock: any suggestions?

rozmansi commented:
The MSI packages will install binaries according to the next matrix:

| Windows Architecture | Driver | Native binaries[*] | `openvpnserv2.exe` |
| --- | --- | --- | --- |
| x86 | x86 | x86 | AnyCPU |
| x64 | x64 | x64 | AnyCPU |
| ARM64 | ARM64 | x86 | AnyCPU |

[*] `openvpn.exe`, `openvpn-gui.exe`, `openvpnserv.exe`, required DLLs...

Therefore, we can't just always expect 64-bit OpenVPN installation on 64-bit Windows. We will shoot ourselves in the foot when we release ARM64 version. What we can assume is, no system shall have both 32 and 64-bit versions of binaries installed at once.

We don't have MinGW, OpenSSL and other libraries available for ARM64 yet and will have to use x86 user-space native binaries for the time being. However, we will hit a problem when we go out with an all-ARM64 package: updating the ARM64/x86 _Frankenstein_ installations will require migrating config files from `C:\Program Files (x86)\OpenVPN\config` to `C:\Program Files\OpenVPN\config`, and registry keys from `Wow6432Node` key.

selvanair commented:
> What we can assume is, no system shall have both 32 and 64-bit versions of binaries installed at once.

Okay, let's go with that but leave the current logic of scanning and starting configs based on both normal and Wow6432Node keys. The new version of service will look for autostart_config_dir key (is that the name we settled on?) in both nodes.  Makes it very simple.

We support cases were Wow6432Node key point to a 64 bit installation (exe_path, config_dir etc.) so there could be some corner cases which admins will have to fix manually.

I'm not worried about ARM as there are no existing (NSIS) installations on that platform so we start from a clean slate and the new reg key and config folder will always exist (empty or not). But your point about valid 32 bit installations on 64bit is good to keep in mind.

rozmansi commented:
> 
> 
> Okay, let's go with that but leave the current logic of scanning and starting configs based on both normal and Wow6432Node keys. The new version of service will look for autostart_config_dir key (is that the name we settled on?) in both nodes. Makes it very simple.

`autostart_config_dir` and `C:\ProgramFiles\OpenVPN\config-auto` is OK with me.

selvanair commented:
Agreed.

selvanair commented:
See PR #142 and [PR 10 of OpenVPN/openvpnserv2](https://github.com/OpenVPN/openvpnserv2/pull/10)

rozmansi commented:
> 
> 
> ... So almost certainly some people have installer x86 version of 2.3.x on x86 Windows.

You meant "x86 version on x64 Windows", right? Actually, such setup wouldn't work, unless they installed the x64 TAP driver by some other means.

mattock commented:
@rozmansi exactly, I typoed. It used to be possible to install 32-bit OpenVPN on 64-bit Windows. The tap-windows[6] installer included in both 32-bit and 64-bit OpenVPN NSI packages contained both 32-bit and 64-bit drivers and the correct one was installed based on OS bitness. So userspace could be 32-bit yet the driver would be 64-bit.

selvanair commented:
I think we need not set it up OpenVPNServiceLegacy in 2.5.x releases. When we renamed it in 2.4, the idea was to wait and watch whether anyone missed the original and then remove it at a later date. We can eventually purge it from sources, but to start with not setting up during install would be an easy first step. One less thing to keep track of.

selvanair commented on windows-msi/doc/config/Where are my config files.txt.in:
```diff
@@ -0,0 +1,6 @@
+
+The update process detected that you were using @PRODUCT_NAME@Service
+to start your connections, and moved all your config files from config
+to the config-auto directory.
```
Telling the user about change of location is good, but should we "move" the files or "copy". I thought "copy" was less destructive and good enough?

rozmansi replied:
Perhaps I'm too eager to clean things. :)

Being a copy or move, to be triggered, it has to be bound to an MSI component (file or registry value). Therefore, I introduced the `Where are my config files.txt` to serve as an anchor, and it all turned out quite naturally.

@mattock, we need the third judge to decide "copy or move" config files from `C:\Program Files\OpenVPN\config` to `C:\Program Files\OpenVPN\config-auto` when updating an OpenVPN installation that is using `openvpnserv2.exe`?

mattock replied:
@rozmansi there will be confusion no matter what we do:

- Copy: "why don't my changes to config <n> in config directory stick?"
- Move: "where did my configs go?"

Personally I think the latter is the smaller evil. It won't be that difficult for a user to figure out that OpenVPN is working -> config files must be somewhere -> go looking -> find config-auto -> look inside.

mattock commented:
I finally got into testing this. I've automated the build environment setup in Vagrant. The build VM runs Ubuntu 18.04 and Windows Server 2016 is used for packaging. The results are in [this openvpn-vagrant PR](https://github.com/OpenVPN/openvpn-vagrant/pull/7). The primary annoyance right now is that @rozmansi 's patches are not yet in "master" in openvpn or openvpn-build. The whole build process can be seen from the changes to README.md in the above PR.

Once correct codebases are used then building MSI packages _almost_ works, but fails when trying to package openvpnserv2.exe:

```
Microsoft (R) Windows Script Host Version 5.812
Copyright (C) Microsoft Corporation. All rights reserved.
--- snip ---
BUILD: tmp\openvpnserv2.exe
GET: https://github-production-release-asset-2e65be.s3.amazonaws.com/162488784/000a3900-03d8-11e9-8c93-8b13894a9abe?X-Am
z-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20181219%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date
=20181219T205420Z&X-Amz-Expires=300&X-Amz-Signature=763433da1daf2a72d287dec78d86c74a0047e3e56be5189a652d2009ce032469&X-A
mz-SignedHeaders=host&actor_id=6889694&response-content-disposition=attachment%3B%20filename%3DOpenVpnServ2.exe&response
-content-type=application%2Foctet-stream >> tmp\openvpnserv2.exe
O:\windows-msi\build.wsf(933, 18) Microsoft JScript runtime error: GET https://github-production-release-asset-2e65be.s3
.amazonaws.com/162488784/000a3900-03d8-11e9-8c93-8b13894a9abe?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJ
YAX4CSVEH53A%2F20181219%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20181219T205420Z&X-Amz-Expires=300&X-Amz-Signature=76
3433da1daf2a72d287dec78d86c74a0047e3e56be5189a652d2009ce032469&X-Amz-SignedHeaders=host&actor_id=6889694&response-conten
t-disposition=attachment%3B%20filename%3DOpenVpnServ2.exe&response-content-type=application%2Foctet-stream failed with status 403 Forbidden.

O:\windows-msi\build.wsf(154, 18) Microsoft JScript runtime error: File not found
```
This probably happens because the URL in version.m4 is inaccessible:
```
dnl OpenVPNServ2.exe binary
dnl TODO: replace-before-flight: http://build.openvpn.net/downloads/releases/openvpnserv2-1.4.0.0.exe                                                        
define([OPENVPNSERV2_URL], [https://github-production-release-asset-2e65be.s3.amazonaws.com/162488784/000a3900-03d8-11e9-8c93-8b13894a9abe?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20181219%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20181219T205420Z&X-Amz-Expires=300&X-Amz-Signature=763433da1daf2a72d287dec78d86c74a0047e3e56be5189a652d2009ce032469&X-Amz-SignedHeaders=host&actor_id=6889694&response-content-disposition=attachment%3B%20filename%3DOpenVpnServ2.exe&response-content-type=application%2Foctet-stream])
```
When I point version.m4 to an URL that works (i.e. openvpnserv1.3.0.0.exe) then I get farther:
```
--- snip ---
RUN: gunzip.exe  -c "sources\easy-rsa-2.3.3_master.tar.gz" | tar.exe  -xC "tmp"
O:\windows-msi\build.wsf(318, 14) ADODB.Stream: File could not be opened.
--- snip ---
```
It seems that Git Bash mostly satisfies the dependencies for tar.exe, gunzip.exe, etc. but not quite. @rozmansi what do you recommend to install to get those tools.

rozmansi commented:
> Once correct codebases are used then building MSI packages _almost_ works, but fails when trying to package openvpnserv2.exe:
> 
> ```
> Microsoft (R) Windows Script Host Version 5.812
> Copyright (C) Microsoft Corporation. All rights reserved.
> --- snip ---
> BUILD: tmp\openvpnserv2.exe
> GET: https://github-production-release-asset-2e65be.s3.amazonaws.com/162488784/000a3900-03d8-11e9-8c93-8b13894a9abe?X-Am
> z-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20181219%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date
> =20181219T205420Z&X-Amz-Expires=300&X-Amz-Signature=763433da1daf2a72d287dec78d86c74a0047e3e56be5189a652d2009ce032469&X-A
> mz-SignedHeaders=host&actor_id=6889694&response-content-disposition=attachment%3B%20filename%3DOpenVpnServ2.exe&response
> -content-type=application%2Foctet-stream >> tmp\openvpnserv2.exe
> O:\windows-msi\build.wsf(933, 18) Microsoft JScript runtime error: GET https://github-production-release-asset-2e65be.s3
> .amazonaws.com/162488784/000a3900-03d8-11e9-8c93-8b13894a9abe?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJ
> YAX4CSVEH53A%2F20181219%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20181219T205420Z&X-Amz-Expires=300&X-Amz-Signature=76
> 3433da1daf2a72d287dec78d86c74a0047e3e56be5189a652d2009ce032469&X-Amz-SignedHeaders=host&actor_id=6889694&response-conten
> t-disposition=attachment%3B%20filename%3DOpenVpnServ2.exe&response-content-type=application%2Foctet-stream failed with status 403 Forbidden.
> 
> O:\windows-msi\build.wsf(154, 18) Microsoft JScript runtime error: File not found
> ```
> 
> This probably happens because the URL in version.m4 is inaccessible:
> 
> ```
> dnl OpenVPNServ2.exe binary
> dnl TODO: replace-before-flight: http://build.openvpn.net/downloads/releases/openvpnserv2-1.4.0.0.exe                                                        
> define([OPENVPNSERV2_URL], [https://github-production-release-asset-2e65be.s3.amazonaws.com/162488784/000a3900-03d8-11e9-8c93-8b13894a9abe?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20181219%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20181219T205420Z&X-Amz-Expires=300&X-Amz-Signature=763433da1daf2a72d287dec78d86c74a0047e3e56be5189a652d2009ce032469&X-Amz-SignedHeaders=host&actor_id=6889694&response-content-disposition=attachment%3B%20filename%3DOpenVpnServ2.exe&response-content-type=application%2Foctet-stream])
> ```
> 
> When I point version.m4 to an URL that works (i.e. openvpnserv1.3.0.0.exe)

Exactly. Not having [Selva's v1.4.0.0](https://github.com/OpenVPN/openvpnserv2/pull/10) compiled and published officially is an issue. Unfortunately, 1.4.0.0 is _required_ in order for MSI package to work correctly. I have compiled and published my own binaries of [openvpnserv2.exe v1.4.0.0](https://github.com/rozmansi/openvpnserv2/releases/tag/v1.4.0.0). Unfortunately, the GitHub is using 30x redirects and the ActiveX object used to download files (`MSXML2.XMLHTTP.6.0`) does not support 30x redirects, so I used target CDN URL to get around this problem quickly. Obviously, the CDN URL changed meanwhile. I have switched to a more recent and powerful ActiveX object to do the downloading.

> then I get farther:
> 
> ```
> --- snip ---
> RUN: gunzip.exe  -c "sources\easy-rsa-2.3.3_master.tar.gz" | tar.exe  -xC "tmp"
> O:\windows-msi\build.wsf(318, 14) ADODB.Stream: File could not be opened.
> --- snip ---
> ```
> 
> It seems that Git Bash mostly satisfies the dependencies for tar.exe, gunzip.exe, etc. but not quite. @rozmansi what do you recommend to install to get those tools.

I was using an ancient distribution of GNU tools for Win32. I have updated the build process to accept Git for Windows packaged utilities. Switching from `gunzip.exe` to `gzip.exe -d` made it happy again.

mattock commented:
I'll take another stab at this soon. There's nothing blocking us/me from publishing openvpnserv2 v1.4.0.0 officially, so I'll look into that as well.

mattock commented:
I finally managed to successfully produce MSI packages. I did not test the packages, as I did not want to play with tapctl.exe signing etc. quite yet. I'd be ok with merging this PR because:

* It has seen a fair amount of review
* The test MSI packages have been tested by a couple of people
* This does not affect our current builds (NSIS etc.) at all

That said, I found out an interesting feature/bug in windows-msi using the "msibuilder" VM (Windows Server 2016 Datacenter edition) in openvpn-vagrant repo. I initially tried to use a Powershell prompt to build. The build started, but stalled at this point:
```
RUN: gzip.exe -d -c "sources\easy-rsa-2.3.3_master.tar.gz" | tar.exe  -xC "tmp"
```
While the build stalls gzip.exe consume tons of CPU:
```
PS> Get-Process|Where-Object { $_.CPU -gt 20 -and $_.ProcessName -eq "gzip" }|Select CPU, ProcessName
CPU ProcessName
--- -----------
222.0625 gzip                                                                                                           
```
The problem is not gzip.exe, but tar.exe. In Powershell tar.exe is (stripped-down/obsolete?) BSD tar:

```
PS C:\> tar.exe --help
tar.exe(bsdtar): manipulate archive files
```
Based on its help and brief testing there's no way to read from STDIN on that particular tar version.

If I run the build in Git Bash the build succeeds, because it is using GNU tar:
```
$ tar.exe --help
Usage: tar [OPTION...] [FILE]...
GNU 'tar' saves many files together into a single tape or disk archive, and can
```
I think it would make sense to document this and/or check the tar.exe version before starting the build. Thoughts?

rozmansi commented:
> I think it would make sense to document this and/or check the tar.exe version before starting the build. Thoughts?

Updating documentation would be easiest and fastest approach. Frankly, I'm not very fond of probing external tools and parsing their output. It might shoot somebody in the foot one day.

I guess there is no compatible command line to make both tar versions work, right?

selvanair commented:
I've no idea how various forms of tar works on Windows but both bsdtar and GNU tar can automatically detect gzipped archives and extract as

`tar -xzf "sources\easy-rsa-2.3.3_master.tar.gz"  -C "tmp"`

No need to pipe the output of gzip.exe -d. Here the -z option to tar is actually not required in (but doesn't hurt) as both bsdtar and gnu tar can automatically detect gzip compression.

If, for some reason, the tar.exe binary is built without zlib support then piping would be required. In that case use ` -f`  option with `-` as the file name to indicate stdin. Like this

`gzip.exe -d -c "sources\easy-rsa-2.3.3_master.tar.gz" | tar.exe  -xzf - -C "tmp"`

Note that -z is probably required in this case.

rozmansi commented:
I shall test and implement @selvanair's suggestion in the following days. @mattock, you should probably wait for this to stabilise before merging this PR.

ikreb7 commented:
Hello,
would it be still possible to add my own config files to the installer, so I could offer my own client for my openvpn server? And are the separated parts still signed?

I do this with the nsis installer, so it would be very awesome if this is still possible.

rozmansi commented:
> Hello,
> would it be still possible to add my own config files to the installer, so I could offer my own client for my openvpn server? And are the separated parts still signed?
> 
> I do this with the nsis installer, so it would be very awesome if this is still possible.

As a matter of fact, this is possible with every MSI already... I recommend a utility called Orca (deployed by Microsoft). Using Orca, you can create an MST transform file in which you add your config files to the Files table. As long as we won't drastically change MSI internal structure, you will be able to reuse the same MST for all OpenVPN MSI versions to come.

I shall describe the exact procedure in the Deployment Manual. Hopefully, before the MSI packaged OpenVPN is released to the general public.

mattock commented:
@rozmansi I suppose using MST transforms quite a bit less effort than rebuilding NSIS installer. Is the MST file outside of the MSI, or inside? As in: does using an MST require resigning the MSI?

rozmansi commented:
> @rozmansi I suppose using MST transforms seem quite a bit less effort than rebuilding NSIS installer. Is the MST file outside of the MSI, or inside? As in: does using an MST require resigning the MSI?

MST is outside of MSI. Imagine it as a diff database on top of MSI database. The MSIs remain intact and thus do not require resigning.

Installation is then:
`msiexec.exe /i OpenVPN-2.5.x-amd64.msi TRANSFORMS="<path to MST1>;<path to MST2>;..."`

MST needs to apply on top of MSI without conflicts and can be reused over a range of product releases, as long as it applies on top of each release MSI without conflicts. Usually, the same MST can be used for x86 and amd64 MSIs.

We normaly use it to override annoying desktop shortcut creation from a particular MSI and other various customizations when mass deploying using Group Policy.

In OpenVPN case admins will use MSTs to adjust default feature levels according to their preferences, default install folder, OpenVPN GUI desktop shortcut and start-on-logon, etc.

When deployed over Group Policy, you add MSI for the product to be installed and list all the MSTs that needs to be applied.

Group Policy itself already has mechanisms to copy _.ovpn_ files. When deployed individually, I can describe exact procedure, how to prepare MST to copy extra files, and optionally update _boot.wsf_ and make back a SFX comprising _boot.wsf_, original MSIs, MST(s), and extra files. SFX doesn't require digital signing when not distributed over the internet: it runs non-elevated.

If it doesn't need to be fancy and self-contained in a SFX, a simple batch file can do too:
```
set plat=%PROCESSOR_ARCHITEW6432%
if "%plat%"=="" set plat=%PROCESSOR_ARCHITECTURE%
if "%PROCESSOR_ARCHITECTURE%"=="" set plat=x86
set msi_name=OpenVPN-2.5.x-%plat%.msi
set log_name=%TEMP%\OpenVPN-install.log
set install_dir=C:\Program Files\OpenVPN
msiexec.exe /i "%msi_name%" /l* "%log_name%" INSTALLDIR="%install_dir%"
copy my.ovpn "%install_dir%" /y
```

rozmansi commented:
> I've no idea how various forms of tar works on Windows but both bsdtar and GNU tar can automatically detect gzipped archives and extract as
> 
> `tar -xzf "sources\easy-rsa-2.3.3_master.tar.gz" -C "tmp"`
> 
> No need to pipe the output of gzip.exe -d. Here the -z option to tar is actually not required (but doesn't hurt) as both bsdtar and gnu tar can automatically detect gzip compression.

To be honest, I was using an ancient version of GNU tar.exe without zlib support. Hence, the implementation with piping.

> If, for some reason, the tar.exe binary is built without zlib support then piping would be required. In that case use ` -f` option with `-` as the file name to indicate stdin. Like this
> 
> `gzip.exe -d -c "sources\easy-rsa-2.3.3_master.tar.gz" | tar.exe -xzf - -C "tmp"`
> 
> Note that -z is probably required in this case.

Actually, -z is not required, even more. It confuses tar.exe to try to unzip already unzipped stream. gzip.exe's output is a TAR stream, not GZiped TAR stream.

I have put a note in requirement-section of the manual that GNU tar.exe is required.

selvanair commented:
On Sat, Mar 30, 2019 at 3:10 PM Simon Rozman <notifications@github.com>
wrote:

>
> If, for some reason, the tar.exe binary is built without zlib support then
> piping would be required. In that case use -f option with - as the file
> name to indicate stdin. Like this
>
> gzip.exe -d -c "sources\easy-rsa-2.3.3_master.tar.gz" | tar.exe -xzf - -C
> "tmp"
>
> Note that -z is probably required in this case.
>
> Actually, -z is not required, even more. It confuses tar.exe to try to
> unzip already unzipped stream. gzip.exe's output is a TAR stream, not
> GZiped TAR stream.
>

Of course -- I don't know what I was thinking. When piping gzipped stream
to "tar -x",  some versions require -z, but in this case we are piping
uncompressed data so no -z should be used at all.. As you figured out.

The last time I checked, tar that comes with Windows 10 is bsd tar built
with zlib, so that would just work without pre-dcompressing by gzip.exe.
But this approach you have ended up with is fine and may be more widely
usable.

Selva

>

rozmansi commented:
Force-push was a result of a rebase on top of new master. Cleaning the dust...

mattock commented:
I was able to build MSI packages that use the tap-windows6 MSM. Installation with default settings on a computer with no OpenVPN installations (=uninstalled NSIS beforehand) worked on these platforms:
* Windows 7 64-bit
* Windows Server 2016 64-bit
* Windows 10 32-bit

It failed on Windows 10 arm64 due to tap-windows6 and wintun issues. I ran out of time before being able to debug that further. It could be a problem in my build, or not. I'll fill in more details later. But the "more important" platforms seem to be ok.

For some reason Windows Server 2016 informed me that the computer needs a restart to complete the installation. Not sure why.

My builds are here:
* https://build.openvpn.net/downloads/temp/msi/

mattock commented:
The MSI installers were produced and "seem to work" just fine.
